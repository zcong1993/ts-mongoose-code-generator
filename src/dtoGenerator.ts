import { Project, SourceFile } from 'ts-morph'
import {
  ParsedType,
  TypeEnum,
  parseSchema
} from '@zcong/mongoose-schema-parser'
import * as mongoose from 'mongoose'
import * as camelcase from 'camelcase'

export interface DtoGeneratorInitOptions {
  file?: SourceFile
  filename?: string
  useInterface?: boolean
}

export class DtoGenerator {
  private file: SourceFile
  private useInterface: boolean
  constructor(opts: DtoGeneratorInitOptions) {
    this.useInterface = opts.useInterface
    if (opts.file) {
      this.file = opts.file
    } else {
      // init
      const project = new Project()
      this.file = project.createSourceFile(
        opts.filename || 'tmp.ts',
        {},
        { overwrite: true }
      )
    }
  }

  generateDtoBySchema(schema: mongoose.Schema, name: string) {
    const parsed = parseSchema(schema)
    this.generateDtoByParsedSchema(parsed, name)
  }

  generateDtoByParsedSchema(parsed: ParsedType, name: string) {
    let declar = this.useInterface
      ? this.file.addInterface({
          name: camelcase(`${name}Dto`, { pascalCase: true }),
          isExported: true
        })
      : this.file.addClass({
          name: camelcase(`${name}Dto`, { pascalCase: true }),
          isExported: true
        })
    Object.keys(parsed).forEach(propKey => {
      const field = parsed[propKey]

      let hasQuestionToken = false
      if (!field.details) {
        hasQuestionToken = true
      } else {
        hasQuestionToken = !field.details.required
      }

      switch (field.type.type) {
        case TypeEnum.Boolean:
        case TypeEnum.String:
        case TypeEnum.Number:
          declar.addProperty({
            name: propKey,
            type: field.type.isArray
              ? `${field.type.type.toLocaleLowerCase()}[]`
              : field.type.type.toLocaleLowerCase(),
            hasQuestionToken
          })
          break
        case TypeEnum.Date:
          declar.addProperty({
            name: propKey,
            type: field.type.isArray ? 'Date[]' : 'Date',
            hasQuestionToken
          })
          break
        case TypeEnum.ObjectId:
          declar.addProperty({
            name: propKey,
            type: field.type.isArray ? 'string[]' : 'string',
            hasQuestionToken
          })
          break
        case TypeEnum.Schema:
          const subTypeName = camelcase(`${name}-${propKey}Sub`, {
            pascalCase: true
          })
          this.generateDtoByParsedSchema(field.schema, subTypeName)
          declar.addProperty({
            name: propKey,
            type: field.type.isArray
              ? `${subTypeName}Dto[]`
              : `${subTypeName}Dto`,
            hasQuestionToken
          })
          break
        default:
          break
      }
    })
  }

  getGeneratedCode() {
    return this.file.getFullText()
  }

  getFile() {
    return this.file
  }
}
