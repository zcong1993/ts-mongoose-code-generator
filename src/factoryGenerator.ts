import { inspect } from 'util'
import {
  parseSchema,
  ParsedType,
  TypeEnum,
  ParsedField
} from '@zcong/mongoose-schema-parser'
import { Project, SourceFile, IndentationText, QuoteKind } from 'ts-morph'
import * as mongoose from 'mongoose'
import * as camelcase from 'camelcase'

export interface CustomField {
  type: TypeEnum
  value: any
  fieldName: string
}

export interface FactoryGeneratorOptions {
  file?: SourceFile
  filename?: string
  dtoFilePath: string
  customFields?: CustomField[]
}

export enum helperFlags {
  shouldAddObjectIdFunc,
  shouldAddEnumPickFunc
}

export class FactoryGenerator {
  private file: SourceFile
  private opts: Readonly<FactoryGeneratorOptions>

  // maybe can migrate to options
  private stringRandom: string = 'faker.lorem.word()'
  private numberRandom: string = 'faker.random.number()'
  private booleanRandom: string = 'faker.random.boolean()'
  private dateRandom: string = 'faker.date.recent()'

  private shouldAddFuncSet: Set<helperFlags> = new Set()
  private isSettedFuncSet: Set<helperFlags> = new Set()

  constructor(opts: FactoryGeneratorOptions) {
    if (opts.file) {
      this.file = opts.file
    } else {
      // init
      const project = new Project({
        manipulationSettings: {
          indentationText: IndentationText.TwoSpaces,
          quoteKind: QuoteKind.Single
        }
      })
      this.file = project.createSourceFile(
        opts.filename || 'tmp.ts',
        {},
        { overwrite: true }
      )
    }
    this.opts = opts

    this.init()
    this.initFileImports()
  }

  generateFactoriesBySchema(schema: mongoose.Schema, name: string) {
    const parsed = parseSchema(schema)
    this.generateFactoriesByParsedSchema(parsed, name)
    this.afterGenerate()
  }

  generateFactoriesByParsedSchema(parsed: ParsedType, name: string) {
    const dtoName = camelcase(`${name}Dto`, { pascalCase: true })

    const funcName = camelcase(`${name}-Factory`)
    const func = this.file.addFunction({
      name: funcName,
      parameters: [
        {
          name: 'initial',
          type: `Partial<dto.${dtoName}>`,
          initializer: '{}'
        }
      ],
      returnType: `dto.${dtoName}`,
      isExported: true
    })

    func.setBodyText(writer => {
      writer.writeLine(`const mock: dto.${dtoName} = {`)

      Object.keys(parsed).forEach(propKey => {
        const field = parsed[propKey]
        const fd = this.getFieldByType(field, propKey, name)
        if (fd) {
          writer.writeLine(`  ${propKey}: ${fd},`)
        }
      })

      writer.writeLine('}').writeLine('return { ...mock, ...initial }')
    })
  }

  getGeneratedCode() {
    return this.file.getFullText()
  }

  getFile() {
    return this.file
  }

  private getFieldByType(
    field: ParsedField,
    propKey: string,
    name: string
  ): string | undefined {
    const isArray = field.type.isArray
    // custom default field name
    if (this.opts.customFields && this.opts.customFields.length > 0) {
      const cfs = this.opts.customFields.filter(
        cf => cf.fieldName === propKey && cf.type === field.type.type
      )

      if (cfs.length > 0) {
        return this.arrayWrap(inspect(cfs[0].value), isArray)
      }
    }

    if (field.details && field.details.enum) {
      this.shouldAddFuncSet.add(helperFlags.shouldAddEnumPickFunc)
      return this.arrayWrap(
        `enumPick(${inspect(field.details.enum)})()`,
        isArray
      )
    }

    let factoryFunc: string

    switch (field.type.type) {
      case TypeEnum.Boolean:
        factoryFunc = this.booleanRandom
        break
      case TypeEnum.String: {
        factoryFunc = this.stringRandom
        break
      }
      case TypeEnum.Number: {
        factoryFunc = this.numberRandom
        break
      }
      case TypeEnum.Date:
        factoryFunc = this.dateRandom
        break
      case TypeEnum.ObjectId:
        this.shouldAddFuncSet.add(helperFlags.shouldAddObjectIdFunc)
        factoryFunc = 'mongoObjectId()'
        break
      case TypeEnum.Schema:
        const subTypeName = camelcase(`${name}-${propKey}Sub`, {
          pascalCase: true
        })
        this.generateFactoriesByParsedSchema(field.schema, subTypeName)
        factoryFunc = `${camelcase(`${subTypeName}-Factory`)}()`
        break
      default:
        return undefined
    }

    return this.arrayWrap(factoryFunc, isArray)
  }

  private arrayWrap(origin: string, isArray: boolean) {
    return isArray ? `[${origin}]` : origin
  }

  private addObjectIdFunc() {
    const func = this.file.addFunction({
      name: 'mongoObjectId',
      returnType: 'string'
    })

    func.setBodyText(writer => {
      writer
        .writeLine(
          'const timestamp = (new Date().getTime() / 1000 | 0).toString(16)'
        )
        .writeLine(
          `return timestamp + 'xxxxxxxxxxxxxxxx'.replace(/[x]/g, () => (Math.random() * 16 | 0).toString(16)).toLowerCase()`
        )
    })
  }

  private addEnumPickFunc() {
    const func = this.file.addFunction({
      name: 'enumPick'
    })

    func.addTypeParameter({
      name: 'T',
      default: 'any'
    })

    func.addParameter({
      name: 'arr',
      type: 'T[]'
    })

    func.setReturnType('() => T')

    func.setBodyText(writer => {
      writer
        .writeLine('function randomPick(): T {')
        .writeLine('  const randomI = Math.floor((Math.random() * arr.length))')
        .writeLine('  return arr[randomI]')
        .writeLine('}')
        .writeLine('return randomPick')
    })
  }

  private afterGenerate() {
    if (
      this.shouldAddFuncSet.has(helperFlags.shouldAddObjectIdFunc) &&
      !this.isSettedFuncSet.has(helperFlags.shouldAddObjectIdFunc)
    ) {
      this.addObjectIdFunc()
      this.isSettedFuncSet.add(helperFlags.shouldAddObjectIdFunc)
    }

    if (
      this.shouldAddFuncSet.has(helperFlags.shouldAddEnumPickFunc) &&
      !this.isSettedFuncSet.has(helperFlags.shouldAddEnumPickFunc)
    ) {
      this.addEnumPickFunc()
      this.isSettedFuncSet.add(helperFlags.shouldAddEnumPickFunc)
    }
  }

  private init() {
    this.file.removeText()
  }

  private initFileImports() {
    this.file.addImportDeclarations([
      {
        moduleSpecifier: 'faker',
        defaultImport: '* as faker'
      },
      {
        moduleSpecifier: this.opts.dtoFilePath,
        defaultImport: '* as dto'
      }
    ])
  }
}
