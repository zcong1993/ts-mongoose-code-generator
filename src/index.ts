export * from './dtoGenerator'
export * from './factoryGenerator'
