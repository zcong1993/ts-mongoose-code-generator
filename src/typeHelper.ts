import { Types } from 'mongoose'

export type PopulateType<T> = NonPopulateModelType | T
export type ArrayPopulateType<T> =
  | Array<string>
  | Array<Types.ObjectId>
  | Array<T>
export type NonPopulateModelType = string | Types.ObjectId
export type PopulateModelType<T> = T extends NonPopulateModelType ? never : T
export type ArrayPopulateModelType<T> = T extends Array<NonPopulateModelType>
  ? never
  : T

export function forcePopulateModelType<T>(
  t: PopulateType<T>
): t is PopulateModelType<T>
export function forcePopulateModelType<T>(
  t: ArrayPopulateType<T>
): t is ArrayPopulateModelType<Array<T>>
export function forcePopulateModelType(_: any): boolean {
  return true
}

export function forceStringType<T>(t: PopulateType<T>): t is string
export function forceStringType<T>(t: ArrayPopulateType<T>): t is Array<string>
export function forceStringType(_: any): boolean {
  return true
}

export function forceObjectIDType<T>(t: PopulateType<T>): t is Types.ObjectId
export function forceObjectIDType<T>(
  t: ArrayPopulateType<T>
): t is Array<Types.ObjectId>
export function forceObjectIDType(_: any): boolean {
  return true
}
