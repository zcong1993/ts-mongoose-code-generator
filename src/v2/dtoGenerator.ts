import { Project, SourceFile, IndentationText, QuoteKind } from 'ts-morph'
import { ParsedType, parseSchema } from '@zcong/mongoose-schema-parser/dist/v2'
import * as mongoose from 'mongoose'
import * as camelcase from 'camelcase'

export interface DtoGeneratorInitOptions {
  file?: SourceFile
  filename?: string
  useInterface?: boolean
}

export class DtoGenerator {
  private file: SourceFile
  private useInterface: boolean
  private mongoseImports: Set<string> = new Set()
  constructor(opts: DtoGeneratorInitOptions) {
    this.useInterface = opts.useInterface
    if (opts.file) {
      this.file = opts.file
    } else {
      // init
      const project = new Project({
        manipulationSettings: {
          indentationText: IndentationText.TwoSpaces,
          quoteKind: QuoteKind.Single,
        },
      })
      this.file = project.createSourceFile(
        opts.filename || 'tmp.ts',
        {},
        { overwrite: true }
      )
    }

    this.init()
  }

  generateDtoBySchema(schema: mongoose.Schema, name: string) {
    const parsed = parseSchema(schema)
    this.generateDtoByParsedSchema(parsed, name, false)
  }

  generateDtoByParsedSchema(parsed: ParsedType, name: string, isSub?: boolean) {
    const declar = this.useInterface
      ? this.file.addInterface({
          name: camelcase(`${name}Dto`, { pascalCase: true }),
          isExported: true,
          extends: isSub ? [] : ['Document'],
        })
      : this.file.addClass({
          name: camelcase(`${name}Dto`, { pascalCase: true }),
          isExported: true,
          extends: isSub ? null : 'Document',
        })

    this.mongoseImports.add('Document')

    Object.keys(parsed).forEach((propKey) => {
      const field = parsed[propKey]

      let hasQuestionToken = false
      if (!field.options) {
        hasQuestionToken = true
      } else {
        hasQuestionToken = !field.options.required
      }

      const isArray = field.type.isArray

      switch (field.type.type) {
        case 'Boolean':
        case 'String':
        case 'Number': {
          const t =
            field.type.type === 'String' && field.type.enumValues
              ? field.type.enumValues.map((s) => `'${s}'`).join(' | ')
              : field.type.type.toLocaleLowerCase()
          declar.addProperty({
            hasQuestionToken,
            name: propKey,
            type: this.arrayWrap(t, isArray),
          })
          break
        }
        case 'Date': {
          declar.addProperty({
            hasQuestionToken,
            name: propKey,
            type: this.arrayWrap('Date', isArray),
          })
          break
        }
        case 'ObjectID': {
          this.mongoseImports.add('Types')
          let type: string
          // handle ref type
          if (field.options && field.options.ref) {
            type = this.arrayWrapOr(
              ['string', 'Types.ObjectId', `${field.options.ref}Dto`],
              isArray
            )
            this.mongoseImports.add('Document')
          } else {
            type = this.arrayWrapOr(['string', 'Types.ObjectId'], isArray)
          }
          declar.addProperty({
            type,
            hasQuestionToken,
            name: propKey,
          })
          break
        }
        case 'Schema':
        case 'Embedded': {
          const subTypeName = camelcase(`${name}-${propKey}Sub`, {
            pascalCase: true,
          })
          this.generateDtoByParsedSchema(field.schema, subTypeName, true)
          declar.addProperty({
            hasQuestionToken,
            name: propKey,
            type: this.arrayWrap(`${subTypeName}Dto`, isArray),
          })
          break
        }
        default:
          break
      }
    })
  }

  getGeneratedCode() {
    this.handleMongooseImports()
    return this.file.getFullText()
  }

  getFile() {
    this.handleMongooseImports()
    return this.file
  }

  private init() {
    this.file.removeText()
  }

  private arrayWrap(origin: string, isArray: boolean): string {
    return isArray ? `Array<${origin}>` : origin
  }

  private arrayWrapOr(types: string[], isArray: boolean) {
    if (!isArray) {
      return types.join(' | ')
    }
    return types.map((t) => `Array<${t}>`).join(' | ')
  }

  private handleMongooseImports() {
    if (this.mongoseImports.size > 0) {
      this.file.addImportDeclaration({
        moduleSpecifier: 'mongoose',
        namedImports: [...this.mongoseImports].map((name) => ({ name })),
      })
      this.mongoseImports.clear()
    }
  }
}
