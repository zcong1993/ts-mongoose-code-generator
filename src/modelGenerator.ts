import { Project, SourceFile, IndentationText, QuoteKind } from 'ts-morph'
import { ParsedType, parseSchema } from '@zcong/mongoose-schema-parser/dist/v2'
import * as mongoose from 'mongoose'
import * as camelcase from 'camelcase'

export interface ModelGeneratorInitOptions {
  file?: SourceFile
  filename?: string
  useInterface?: boolean
  stringEnumUseUnionType?: boolean
}

export class ModelGenerator {
  private readonly opts: ModelGeneratorInitOptions
  private file: SourceFile
  private useInterface: boolean
  private mongoseImports: Set<string> = new Set()
  constructor(opts: ModelGeneratorInitOptions) {
    this.opts = opts
    this.useInterface = opts.useInterface
    /* istanbul ignore next */
    if (opts.file) {
      this.file = opts.file
    } else {
      // init
      const project = new Project({
        manipulationSettings: {
          indentationText: IndentationText.TwoSpaces,
          quoteKind: QuoteKind.Single,
        },
      })
      this.file = project.createSourceFile(
        opts.filename || 'tmp.ts',
        `// This file is created by @zcong/ts-mongoose-code-generator`,
        { overwrite: true }
      )
    }
  }

  generateModelBySchema(schema: mongoose.Schema, name: string) {
    const parsed = parseSchema(schema)
    this.generateModelByParsedSchema(parsed, name)
  }

  generateModelByParsedSchema(parsed: ParsedType, name: string) {
    const declar = this.useInterface
      ? this.file.addInterface({
          name: camelcase(`${name}Model`, { pascalCase: true }),
          isExported: true,
        })
      : this.file.addClass({
          name: camelcase(`${name}Model`, { pascalCase: true }),
          isExported: true,
        })

    Object.keys(parsed).forEach((propKey) => {
      const field = parsed[propKey]

      let hasQuestionToken = false
      if (!field.options) {
        hasQuestionToken = true
      } else {
        hasQuestionToken = !field.options.required
      }

      const isArray = field.type.isArray

      switch (field.type.type) {
        case 'Boolean':
        case 'String':
        case 'Number': {
          const t =
            field.type.type === 'String' &&
            field.type.enumValues &&
            this.opts.stringEnumUseUnionType
              ? field.type.enumValues.map((s: string) => `'${s}'`).join(' | ')
              : field.type.type.toLocaleLowerCase()
          declar.addProperty({
            hasQuestionToken,
            name: propKey,
            type: this.arrayWrap(t, isArray),
          })
          break
        }
        case 'Date': {
          declar.addProperty({
            hasQuestionToken,
            name: propKey,
            type: this.arrayWrap('Date', isArray),
          })
          break
        }
        case 'ObjectID': {
          this.mongoseImports.add('Types')
          let type: string
          // handle ref type
          if (field.options && field.options.ref) {
            type = this.arrayWrapOr(
              [
                'string',
                'Types.ObjectId',
                `(${field.options.ref}Model & Document)`,
              ],
              isArray
            )
            this.mongoseImports.add('Document')
          } else {
            type = this.arrayWrapOr(['string', 'Types.ObjectId'], isArray)
          }
          declar.addProperty({
            type,
            hasQuestionToken,
            name: propKey,
          })
          break
        }
        case 'Schema':
        case 'Embedded': {
          const subTypeName = camelcase(`${name}-${propKey}Sub`, {
            pascalCase: true,
          })
          this.generateModelByParsedSchema(field.schema, subTypeName)
          declar.addProperty({
            hasQuestionToken,
            name: propKey,
            type: this.arrayWrap(`${subTypeName}Model`, isArray),
          })
          break
        }
        case 'Mixed': {
          declar.addProperty({
            hasQuestionToken,
            name: propKey,
            type: this.arrayWrap('any', isArray),
          })
          break
        }
        case 'Buffer': {
          declar.addProperty({
            hasQuestionToken,
            name: propKey,
            type: this.arrayWrap('Buffer', isArray),
          })
          break
        }
        case 'Map': {
          declar.addProperty({
            hasQuestionToken,
            name: propKey,
            type: this.arrayWrap('any', isArray), // todo: use any
          })
          break
        }
        case 'Decimal128': {
          declar.addProperty({
            hasQuestionToken,
            name: propKey,
            type: this.arrayWrap('Types.Decimal128', isArray),
          })
          this.mongoseImports.add('Types')
          break
        }
        /* istanbul ignore next */
        default:
          /* istanbul ignore next */
          declar.addProperty({
            hasQuestionToken,
            name: propKey,
            type: this.arrayWrap('any', isArray),
          })
          /* istanbul ignore next */
          console.log(
            `[@zcong/ts-mongoose-code-generator] unhandled, propKey: ${propKey}, type: ${
              field.type.type
            }, options: ${JSON.stringify(field)}`
          )
      }
    })
  }

  getGeneratedCode() {
    this.handleMongooseImports()
    return this.file.getFullText()
  }

  getFile() {
    this.handleMongooseImports()
    return this.file
  }

  private arrayWrap(origin: string, isArray: boolean): string {
    return isArray ? `Array<${origin}>` : origin
  }

  private arrayWrapOr(types: string[], isArray: boolean) {
    if (!isArray) {
      return types.join(' | ')
    }
    return types.map((t) => `Array<${t}>`).join(' | ')
  }

  private handleMongooseImports() {
    if (this.mongoseImports.size > 0) {
      this.file.addImportDeclaration({
        moduleSpecifier: 'mongoose',
        namedImports: [...this.mongoseImports].map((name) => ({ name })),
      })
      this.mongoseImports.clear()
    }
  }
}
