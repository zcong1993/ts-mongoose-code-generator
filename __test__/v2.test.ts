import * as fs from 'fs'
import { Schema } from 'mongoose'
import { DtoGenerator } from '../src/v2'

const ObjectId = Schema.Types.ObjectId

const nestedSchema = new Schema({
  nestedName: String,
})

const refSchema = new Schema(
  {
    refName: String,
  },
  { timestamps: true }
)

const testSchema = new Schema(
  {
    id: ObjectId,
    name: String,
    age: {
      type: Number,
    },
    requiredName: {
      type: String,
      required: true,
    },
    enumString: {
      type: String,
      enum: ['test1', 'test2'],
    },
    date: Date,
    bool: Boolean,
    nested: nestedSchema,
    nestedArr: [nestedSchema],
    nestedArr2: [
      {
        type: nestedSchema,
      },
    ],
    simpleArr: [String],
    simpleArr2: [
      {
        type: String,
      },
    ],
    directNested: {
      name: String,
      age: Number,
    },
    directNestedArr: [
      {
        name: String,
        age: Number,
      },
    ],
    ref: {
      type: ObjectId,
      ref: 'Ref',
    },
    refs: [
      {
        type: ObjectId,
        ref: 'Ref',
      },
    ],
    withTypeField: [
      {
        type: {
          type: String,
        },
        other: String,
      },
    ],
  },
  { timestamps: { createdAt: 'created_at', updatedAt: 'updated_at' } }
)

it('v2 dtoGenerator should work well', () => {
  const dtoGen = new DtoGenerator({ filename: `${__dirname}/dtoGenv2.ts` })
  dtoGen.generateDtoBySchema(testSchema, 'Test')
  dtoGen.generateDtoBySchema(refSchema, 'Ref')
  dtoGen.getFile().saveSync()
  const content = fs.readFileSync(`${__dirname}/dtoGenv2.ts`, 'utf8')
  const generated = dtoGen.getGeneratedCode()
  expect(content).toEqual(generated)
  expect(generated).toMatchSnapshot()
  fs.unlinkSync(`${__dirname}/dtoGenv2.ts`)
})

it('v2 dtoGenerator use interface should work well', () => {
  const dtoGen = new DtoGenerator({
    filename: `${__dirname}/dtoGenv2.ts`,
    useInterface: true,
  })
  dtoGen.generateDtoBySchema(testSchema, 'Test')
  dtoGen.generateDtoBySchema(refSchema, 'Ref')
  dtoGen.getFile().saveSync()
  const content = fs.readFileSync(`${__dirname}/dtoGenv2.ts`, 'utf8')
  const generated = dtoGen.getGeneratedCode()
  expect(content).toEqual(generated)
  expect(generated).toMatchSnapshot()
  fs.unlinkSync(`${__dirname}/dtoGenv2.ts`)
})
