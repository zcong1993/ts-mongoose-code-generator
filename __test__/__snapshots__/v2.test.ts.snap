// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`v2 dtoGenerator should work well 1`] = `
"import { Document, Types } from 'mongoose';

export class TestDto extends Document {
  id?: string | Types.ObjectId;
  name?: string;
  age?: number;
  requiredName: string;
  enumString?: 'test1' | 'test2';
  date?: Date;
  bool?: boolean;
  nested?: TestNestedSubDto;
  nestedArr?: Array<TestNestedArrSubDto>;
  nestedArr2?: Array<TestNestedArr2SubDto>;
  simpleArr?: Array<string>;
  simpleArr2?: Array<string>;
  directNestedArr?: Array<TestDirectNestedArrSubDto>;
  ref?: string | Types.ObjectId | RefDto;
  refs?: Array<string> | Array<Types.ObjectId> | Array<RefDto>;
  withTypeField?: Array<TestWithTypeFieldSubDto>;
  _id?: string | Types.ObjectId;
  updated_at?: Date;
  created_at?: Date;
  directNested?: TestDirectNestedSubDto;
}

export class TestNestedSubDto {
  nestedName?: string;
  _id?: string | Types.ObjectId;
}

export class TestNestedArrSubDto {
  nestedName?: string;
  _id?: string | Types.ObjectId;
}

export class TestNestedArr2SubDto {
  nestedName?: string;
  _id?: string | Types.ObjectId;
}

export class TestDirectNestedArrSubDto {
  name?: string;
  age?: number;
  _id?: string | Types.ObjectId;
}

export class TestWithTypeFieldSubDto {
  type?: string;
  other?: string;
  _id?: string | Types.ObjectId;
}

export class TestDirectNestedSubDto {
  name?: string;
  age?: number;
}

export class RefDto extends Document {
  refName?: string;
  _id?: string | Types.ObjectId;
  updatedAt?: Date;
  createdAt?: Date;
}
"
`;

exports[`v2 dtoGenerator use interface should work well 1`] = `
"import { Document, Types } from 'mongoose';

export interface TestDto extends Document {
  id?: string | Types.ObjectId;
  name?: string;
  age?: number;
  requiredName: string;
  enumString?: 'test1' | 'test2';
  date?: Date;
  bool?: boolean;
  nested?: TestNestedSubDto;
  nestedArr?: Array<TestNestedArrSubDto>;
  nestedArr2?: Array<TestNestedArr2SubDto>;
  simpleArr?: Array<string>;
  simpleArr2?: Array<string>;
  directNestedArr?: Array<TestDirectNestedArrSubDto>;
  ref?: string | Types.ObjectId | RefDto;
  refs?: Array<string> | Array<Types.ObjectId> | Array<RefDto>;
  withTypeField?: Array<TestWithTypeFieldSubDto>;
  _id?: string | Types.ObjectId;
  updated_at?: Date;
  created_at?: Date;
  directNested?: TestDirectNestedSubDto;
}

export interface TestNestedSubDto {
  nestedName?: string;
  _id?: string | Types.ObjectId;
}

export interface TestNestedArrSubDto {
  nestedName?: string;
  _id?: string | Types.ObjectId;
}

export interface TestNestedArr2SubDto {
  nestedName?: string;
  _id?: string | Types.ObjectId;
}

export interface TestDirectNestedArrSubDto {
  name?: string;
  age?: number;
  _id?: string | Types.ObjectId;
}

export interface TestWithTypeFieldSubDto {
  type?: string;
  other?: string;
  _id?: string | Types.ObjectId;
}

export interface TestDirectNestedSubDto {
  name?: string;
  age?: number;
}

export interface RefDto extends Document {
  refName?: string;
  _id?: string | Types.ObjectId;
  updatedAt?: Date;
  createdAt?: Date;
}
"
`;
